### PR TITLE
Narrow tagged unions based on count() with array-size

### PR DESCRIPTION
**Pull request by staabm**
as discussed in https://github.com/phpstan/phpstan/discussions/11473 I put the logic arround the [list to constant-array logic](https://github.com/phpstan/phpstan-src/pull/1817)

### CODE REVIEW

File: src/Analyser/TypeSpecifier.php [hunk 1]
```diff
@@ -984,29 +984,59 @@ private function specifyTypesForConstantBinaryExpression(
...
+			) {
+				$result = [];
+				foreach ($argType->getTypes() as $innerType) {
+					if (!$innerType->isConstantArray()->yes()) {
+						$result = null;
+						break;
+					}
+
+					$arraySize = $innerType->getArraySize();
+					if (!$arraySize instanceof ConstantIntegerType) {
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php]: Just got an idea on how this can be improved. Will look into the PR again later

Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: hmm I can't get the idea pass all tests..

```diff
diff --git a/src/Analyser/TypeSpecifier.php b/src/Analyser/TypeSpecifier.php
index 8ae29d00e..b95074a2d 100644
--- a/src/Analyser/TypeSpecifier.php
+++ b/src/Analyser/TypeSpecifier.php
@@ -1013,15 +1013,18 @@ class TypeSpecifier
 					}
 
 					$arraySize = $innerType->getArraySize();
-					if (!$arraySize instanceof ConstantIntegerType) {
+					if (
+						$arraySize instanceof ConstantIntegerType
+						|| $arraySize instanceof IntegerRangeType
+					) {
+						if (!$arraySize->isSuperTypeOf($constantType)->yes()) {
+							continue;
+						}
+					} else {
 						$result = null;
 						break;
 					}
 
-					if (!$constantType->isSuperTypeOf($arraySize)->yes()) {
-						continue;
-					}
-
 					$result[] = $innerType;
 				}
 
diff --git a/tests/PHPStan/Analyser/nsrt/narrow-tagged-union.php b/tests/PHPStan/Analyser/nsrt/narrow-tagged-union.php
index 5ccb14560..c6664b605 100644
--- a/tests/PHPStan/Analyser/nsrt/narrow-tagged-union.php
+++ b/tests/PHPStan/Analyser/nsrt/narrow-tagged-union.php
@@ -79,6 +79,45 @@ class HelloWorld
 		assertType("array{array<int>, numeric-string}|array{string, '', non-empty-string}", $arr);
 	}
 
+	/** @param array{0: int, 1: int, 2: int, 3?: int}|array{string, ''} $arr */
+	public function optKeysInConstArray(array $arr): void
+	{
+		if (count($arr) === 1) {
+			assertType("*NEVER*", $arr);
+		} else {
+			assertType("array{0: int, 1: int, 2: int, 3?: int}|array{string, ''}", $arr);
+		}
+		assertType("array{0: int, 1: int, 2: int, 3?: int}|array{string, ''}", $arr);
+
+		if (count($arr) === 2) {
+			assertType("array{string, ''}", $arr);
+		} else {
+			assertType("array{0: int, 1: int, 2: int, 3?: int}", $arr);
+		}
+		assertType("array{0: int, 1: int, 2: int, 3?: int}|array{string, ''}", $arr);
+
+		if (count($arr) === 3) {
+			assertType("array{0: int, 1: int, 2: int, 3?: int}", $arr);
+		} else {
+			assertType("array{string, ''}", $arr);
+		}
+		assertType("array{0: int, 1: int, 2: int, 3?: int}|array{string, ''}", $arr);
+
+		if (count($arr) === 4) {
+			assertType("array{0: int, 1: int, 2: int, 3?: int}", $arr);
+		} else {
+			assertType("array{string, ''}", $arr);
+		}
+		assertType("array{0: int, 1: int, 2: int, 3?: int}|array{string, ''}", $arr);
+
+		if (count($arr) === 5) {
+			assertType("*NEVER*", $arr);
+		} else {
+			assertType("array{0: int, 1: int, 2: int, 3?: int}|array{string, ''}", $arr);
+		}
+		assertType("array{0: int, 1: int, 2: int, 3?: int}|array{string, ''}", $arr);
+	}
+
 	/** @param array{string, '', non-empty-string}|array<int> $arr */
 	public function mixedArrays(array $arr): void
 	{
```

**Comment by ondrejmirtes** [comment]
I've simplified and fixed the logic: https://github.com/phpstan/phpstan-src/pull/3302/commits/8d16a7ccbf1c0708162de83697e9bac2be49b9b7

Also I issued a bugfix that made it possible: https://github.com/phpstan/phpstan-src/commit/ed6bc0b93499b5c7463ba05aefb1377b9eb9d1e0

**Comment by ondrejmirtes** [comment]
Also the same logic could essentially be executed for `if (count($x) > $constantInt) {` because the same logic holds even if `$arraySize` is IntegerRangeType. I've added these tests but one of the asserts fails:

```diff
diff --git a/tests/PHPStan/Analyser/nsrt/narrow-tagged-union.php b/tests/PHPStan/Analyser/nsrt/narrow-tagged-union.php
index 8ecf3438e..fa2daa399 100644
--- a/tests/PHPStan/Analyser/nsrt/narrow-tagged-union.php
+++ b/tests/PHPStan/Analyser/nsrt/narrow-tagged-union.php
@@ -107,5 +107,41 @@ class HelloWorld
 		}
 		assertType("array{}|array{'xy'}|array{0: 'ab', 1?: 'xy'}", $x);
 	}
+
+	public function arrayGreatherThan(): void
+	{
+		$x = [];
+		if (rand(0,1)) {
+			$x[] = 'ab';
+		}
+		if (rand(0,1)) {
+			$x[] = 'xy';
+		}
+
+		if (count($x) > 0) {
+			assertType("array{'xy'}|array{0: 'ab', 1?: 'xy'}", $x);
+		} else {
+			assertType("array{}", $x);
+		}
+		assertType("array{}|array{'xy'}|array{0: 'ab', 1?: 'xy'}", $x);
+	}
+
+	public function arrayGreatherThan2(): void
+	{
+		$x = [];
+		if (rand(0,1)) {
+			$x[] = 'ab';
+		}
+		if (rand(0,1)) {
+			$x[] = 'xy';
+		}
+
+		if (count($x) > 1) {
+			assertType("array{0: 'ab', 1?: 'xy'}", $x);
+		} else {
+			assertType("array{}|array{'xy'}|array{0: 'ab', 1?: 'xy'}", $x);
+		}
+		assertType("array{}|array{'xy'}|array{0: 'ab', 1?: 'xy'}", $x);
+	}
 }
 
```

We should be able to reuse the same logic for `} elseif ($expr instanceof Node\Expr\BinaryOp\Smaller || $expr instanceof Node\Expr\BinaryOp\SmallerOrEqual) {` branch in TypeSpecifier.

**Comment by ondrejmirtes** [comment]
I'm going to open a separate issue about it.

**Comment by ondrejmirtes** [comment]
Thank you.

**Comment by ondrejmirtes** [comment]
Here. https://github.com/phpstan/phpstan/issues/11480

**Comment by staabm** [comment]
awesome, thanks

**Comment by stof** [comment]
As long as phpstan does not distinguish sealed shapes and unsealed shapes, this change is totally unsafe because phpstan would allow `array('a', 'b', 'c')` to satisfy a type `array{string, string}` (as it accepts arrays that have additional values unspecified in the shape)

**Comment by staabm** [comment]
@stof please report a new issue with a code snippet reproducing the problem you are talking about

**Comment by ondrejmirtes** [comment]
There are many issues about current array unsealed-ness. PHPStan isn't very consistent about it.

**Comment by stof** [comment]
@staabm here you go: https://github.com/phpstan/phpstan/issues/11494

@ondrejmirtes I would suggest avoiding to add narrowings that requires sealedness to be valid until sealedness is actually implemented.

**Pull request merged**